plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java

```diff
@@ -16,27 +16,82 @@
 package org.gradle.testretry.internal.executer.framework;
 
 import org.gradle.api.internal.tasks.testing.TestFramework;
+import org.gradle.api.internal.tasks.testing.filter.DefaultTestFilter;
 import org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestFramework;
 import org.gradle.api.tasks.testing.junitplatform.JUnitPlatformOptions;
-import org.gradle.testretry.internal.executer.TestFilterBuilder;
 import org.gradle.testretry.internal.executer.TestFrameworkTemplate;
 import org.gradle.testretry.internal.executer.TestNames;
 
+import java.lang.reflect.Constructor;
+
+import static org.gradle.testretry.internal.executer.framework.Junit5TestFrameworkStrategy.TestFrameworkProvider.testFrameworkProvider;
+import static org.gradle.testretry.internal.executer.framework.TestFrameworkStrategy.gradleVersionIsAtLeast;
+
 final class Junit5TestFrameworkStrategy extends BaseJunitTestFrameworkStrategy {
 
     @Override
-    public TestFramework createRetrying(TestFrameworkTemplate template, TestNames failedTests) {
-        TestFilterBuilder filters = template.filterBuilder();
-        addFilters(filters, template.testsReader, failedTests, false);
-        JUnitPlatformTestFramework newFramework = new JUnitPlatformTestFramework(filters.build());
-        copyTestOptions((JUnitPlatformOptions) template.task.getTestFramework().getOptions(), newFramework.getOptions());
-        return newFramework;
+    public TestFramework createRetrying(TestFrameworkTemplate template, TestFramework testFramework, TestNames failedTests) {
+        DefaultTestFilter failedTestsFilter = testFilterFor(failedTests, false, template);
+        return testFrameworkProvider(template, testFramework).testFrameworkFor(failedTestsFilter);
     }
 
-    private void copyTestOptions(JUnitPlatformOptions source, JUnitPlatformOptions target) {
-        target.setIncludeEngines(source.getIncludeEngines());
-        target.setExcludeEngines(source.getExcludeEngines());
-        target.setIncludeTags(source.getIncludeTags());
-        target.setExcludeTags(source.getExcludeTags());
+    interface TestFrameworkProvider {
+        class ProviderForCurrentGradleVersion implements TestFrameworkProvider {
+
+            private final TestFramework testFramework;
+
+            ProviderForCurrentGradleVersion(TestFramework testFramework) {
+                this.testFramework = testFramework;
+            }
+
+            @Override
+            public TestFramework testFrameworkFor(DefaultTestFilter failedTestsFilter) {
+                return testFramework.copyWithFilters(failedTestsFilter);            }
+        }
```

Comment:
❓ Why can't we share this implementation, it is the same across all frameworks.

Reply:
I was also wondering that.

Reply:
Good idea.

DONE (45ed11e2)